server/middleware/setUpValidationMiddleware.ts

```diff
@@ -0,0 +1,47 @@
+import express, { Router, Request, Response, NextFunction } from 'express'
+
+import { FieldValidationError } from '../@types/FieldValidationError'
+
+export type Validator = (body: Record<string, unknown>) => FieldValidationError[]
+
+export const FLASH_KEY_VALIDATION_ERRORS = 'validationErrors'
+export const FLASH_KEY_SUBMITTED_FORM = 'submittedForm'
+
+export function validateRequest(request: Request, validator: Validator) {
+  const errors = validator(request.body)
+  if (errors.length) {
+    request.flash(FLASH_KEY_VALIDATION_ERRORS, errors)
+    request.flash(FLASH_KEY_SUBMITTED_FORM, request.body)
+    throw new ValidationError('A validation error occured')
+  }
```

Comment:
Added a simple validation framework.

Validation works by:
* A custom validator function determines whether one or more errors needs reporting on.
* If so this function sets some flash state and throws a custom exception.
* The `errorHandler` detects this type of error then redirects back to the GET of the original request.
* Some middleware inspects flash state and adds any error message or form input to `res.locals`
* These can then be accessed in `nunjucks` templates

Calling this in a POST handler works nicely for the current use case - we don't have the required metadata to check whether the form data is valid until we're in the handler.


We could extend this to be called from a bit of middleware on a specific POST route for complex forms if we know the schema in advance and it makes sense 

I'd also like to add a custom macro wrapping the error summary bit in a follow up.